File: 09/c/main.c
#include <stdio.h>
#include <assert.h>
#include <string.h>

#define SWAP(x, y) { int SWAP = x; x = y; y = SWAP; };

#define W 100
#define H 100
char tiles[H][W + 1] = {0};

int isLower(int y, int x) {
  char tile = tiles[y][x];

  if (y > 0 && tiles[y - 1][x] <= tile) return 0; // top
  if (x > 0 && tiles[y][x - 1] <= tile) return 0; // left
  if (x < W - 1 && tiles[y][x + 1] <= tile) return 0; // right
  if (y < H - 1 && tiles[y + 1][x] <= tile) return 0; // bottom

  return 1;
}

int getBasin(int y, int x, int checked[W][H]) {
  if (x < 0 || x >= W || y < 0 || y >= H) return 0;
  if (tiles[y][x] == '9') return 0;

  if (checked[y][x]) return 0;

  int count = 1; // itself
  checked[y][x] = 1;

  count += getBasin(y - 1, x, checked);
  count += getBasin(y + 1, x, checked);
  count += getBasin(y, x - 1, checked);
  count += getBasin(y, x + 1, checked);

  return count;
}

int main(int argc, char *argv[]) {
  assert(argc == 2 && "Must provide one argument => input.txt path");

  printf("Input file: %s\n\n", argv[1]);
  FILE *file = fopen(argv[1], "r");

  char str[W];
  int curr = 0;
  while (fscanf(file, "%s", tiles[curr++]) == 1) { }

  // Problem 1
  {
    int answer = 0;

    for (int y = 0; y < H; y++) {
      for (int x = 0; x < W; x++) {
        if (isLower(y, x)) {
          answer += tiles[y][x] + 1 - '0';
        }
      }
    }

    printf("Answer 1: %d\n", answer);
  }


  // Problem 2
  {
    int basins[3] = {0};

    for (int y = 0; y < H; y++) {
      for (int x = 0; x < W; x++) {
        if (isLower(y, x)) {
          int checked[W][H] = {0};
          int basin = getBasin(y, x, checked);

          if (basin > basins[2]) basins[2] = basin;
          if (basins[2] > basins[1]) SWAP(basins[2], basins[1]);
          if (basins[1] > basins[0]) SWAP(basins[1], basins[0]);
        }
      }
    }

    int answer = basins[0] * basins[1] * basins[2];
    printf("Answer 2: %d\n", answer);
  }


  fclose(file);

  return 0;
}

File: 09/js/main.js
const fs = require('fs')
const {join} = require('path')

const file = fs.readFileSync(join(__dirname, '../input.txt'), 'utf8')
const floor = file.split('\n').slice(0, -1).map(line => line.split('').map(n => Number(n)))

const h = floor.length
const w = floor[0].length

function isLower(y, x) {
  const tile = floor[y][x]

  if (y > 0 && floor[y - 1][x] <= tile) return false // top
  if (x > 0 && floor[y][x - 1] <= tile) return false // left
  if (x < w - 1 && floor[y][x + 1] <= tile) return false // right
  if (y < h - 1 && floor[y + 1][x] <= tile) return false // bottom


  return true
}

// Problem 1
{
  let answer = 0

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (isLower(y, x)) {
        answer += floor[y][x] + 1
      }
    }
  }

  console.log(`Answer 1: ${answer}`);
}

function getBasin(y, x, checked = []) {
  if (x < 0 || x >= w || y < 0 || y >= h) return 0
  if (floor[y][x] == 9) return 0

  const key = `${y}_${x}`
  if (checked.includes(key)) return 0

  let count = 1 // itself
  checked.push(key)

  count += getBasin(y - 1, x, checked)
  count += getBasin(y + 1, x, checked)
  count += getBasin(y, x - 1, checked)
  count += getBasin(y, x + 1, checked)

  return count
}

// Problem 2
{
  const basins = []

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (isLower(y, x)) {
        basins.push(getBasin(y, x))
      }
    }
  }

  const sorted = basins.sort((a, b) => b - a)
  const answer = sorted[0] * sorted[1] * sorted[2]
  console.log(`Answer 2: ${answer}`);
}
